src/storage/core.rs

```diff
@@ -1228,14 +1239,46 @@ where
         }
     }
 
-    async fn fsyncdata(&self) -> IOResult<()> {
-        self.safe.read().await.fsyncdata().await
+    pub(crate) async fn fsyncdata(&self) -> IOResult<()> {
+        if self.fsync_in_progress.compare_exchange(false, true, Ordering::Release, Ordering::Acquire).is_err() {
```

Comment:
According to the documentation, first ordering is used both for load and store, so it should be `AcqRel`

```suggestion
        if self.fsync_in_progress.compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire).is_err() {
```